Methods hashCode and equals are overridden

NikolayAbysov opened the pull request.

NikolayAbysov committed:
Methods hashCode and equals are overridden

NikolayAbysov committed:
Equals method switched from instanceOf to getClass

NikolayAbysov committed:
Equals method switched from instanceOf to getClass

RomanDubovskyi commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -27,4 +27,21 @@ public Integer getLength() {
+        int result = 42;
+        result = result * 32 * width.hashCode();
+        result = result * 32 * length.hashCode();
+        result = result * 32 * color.hashCode();
```
check for null before calling "field.hashcode()", cause you can get nullPointer in some cases

RomanDubovskyi commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -27,4 +27,21 @@ public Integer getLength() {
+
+    @Override
+    public boolean equals(Object obj) {
+        return obj.getClass() == this.getClass()
```
there should be 4 checks, you have currently only 2

NikolayAbysov committed:
Issues fixed:
1. Additional checks added to equals method
2. Added null checks in equals method

RomanDubovskyi commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -27,4 +27,32 @@ public Integer getLength() {
+
+    @Override
+    public int hashCode() {
+        if (width != null && length != null && color != null) {
```
each check should be separate, cause you can have situation, where only one field is null and other are ok, so you want to take them into account

RomanDubovskyi commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -27,4 +27,32 @@ public Integer getLength() {
+            return false;
+        }
+        if (obj.getClass().equals(this.getClass())) {
+            return ((Rectangle) obj).getColor().equals(this.color)
```
nice, but better to cast whole object to Rectangle once, than you won't need to cast each field separately

NikolayAbysov committed:
Issues fixed:
1. NUll checks now separate
2. One cast is used
